src/Controller/Component/SecurityComponent.php

```diff
@@ -344,20 +446,99 @@ protected function _validatePost(Controller $controller)
                 }
             }
         }
-        sort($unlocked, SORT_STRING);
         sort($fieldList, SORT_STRING);
         ksort($lockedFields, SORT_STRING);
-
         $fieldList += $lockedFields;
-        $unlocked = implode('|', $unlocked);
-        $hashParts = [
-            $controller->request->here(),
-            serialize($fieldList),
-            $unlocked,
-            Security::salt()
-        ];
-        $check = Security::hash(implode('', $hashParts), 'sha1');
-        return ($token === $check);
+        return $fieldList;
+    }
+
+
+    /**
+     * Get the unlocked string
+     *
+     * @param array $check Data array
+     * @return string
+     */
+    protected function _unlocked(array $check)
+    {
+        return urldecode($check['_Token']['unlocked']);
+    }
+
+    /**
+     * Create a message for humans to understand why Security token is not matching
+     *
+     * @param \Cake\Controller\Controller $controller Instantiating controller
+     * @param array $hashParts Elements used to generate the Token hash
+     * @return array Messages to explain why token is not matching
+     */
+    protected function _debugPostTokenNotMatching(Controller $controller, $hashParts)
+    {
+        $messages = [];
+        $expectedParts = json_decode(urldecode($controller->request->data['_Token']['debug']), true);
+        if (!is_array($expectedParts) || count($expectedParts) !== 3) {
+            return 'Invalid security debug token.';
+        }
+        if ($hashParts[0] !== $expectedParts[0]) {
+            $messages[] = sprintf('URL mismatch in POST data (expected \'%s\' but found \'%s\')', $expectedParts[0], $hashParts[0]);
+        }
+        $expectedFields = $expectedParts[1];
+        $dataFields = unserialize($hashParts[1]);
+        $fieldsMessages = $this->_debugCheckFields(
+            $dataFields,
+            $expectedFields,
+            'Unexpected field \'%s\' in POST data',
+            'Tampered field \'%s\' in POST data (expected value \'%s\' but found \'%s\')',
+            'Missing field \'%s\' in POST data'
+        );
+        $expectedUnlockedFields = Hash::get($expectedParts, 2);
+        $dataUnlockedFields = Hash::get($hashParts, 2) ?: [];
```

Comment:
Yeah, that's great but in this case
    Hash::get($hashParts, 2) // returns '' and we want to convert it to []